core/src/main/scala/org/apache/spark/scheduler/TaskAssigner.scala

```diff
@@ -0,0 +1,233 @@
+
+  // Invoked at the end of resource offering to release internally maintained resources.
+  // Subclass is responsible to release its own private resources.
+  def reset(): Unit = {
+    offer = null
+  }
+}
+
+object TaskAssigner extends Logging {
+  private val roundrobin = classOf[RoundRobinAssigner].getCanonicalName
+  private val packed = classOf[PackedAssigner].getCanonicalName
+  private val balanced = classOf[BalancedAssigner].getCanonicalName
+  private val assignerMap: Map[String, String] =
+    Map("roundrobin" -> roundrobin,
+      "packed" -> packed,
+      "balanced" -> balanced)
+
+  def init(conf: SparkConf): TaskAssigner = {
+    val assignerName = conf.get(config.SPARK_SCHEDULER_TASK_ASSIGNER.key, "roundrobin")
+      .toLowerCase()
+    val className = assignerMap.getOrElse(assignerName, roundrobin)
+    val CPUS_PER_TASK = conf.getInt("spark.task.cpus", 1)
+    val assigner = try {
+      logInfo(s"Constructing an assigner as $className")
+      Utils.classForName(className).getConstructor()
+        .newInstance().asInstanceOf[TaskAssigner]
+    } catch {
+      case _: Throwable =>
+        logInfo(s"$assignerName cannot be constructed, fallback to default $roundrobin.")
+        new RoundRobinAssigner()
+    }
+    assigner.withCpuPerTask(CPUS_PER_TASK)
+    assigner
+  }
+}
+
+/**
+ * Assign the task to workers with available cores in roundrobin manner.
+ */
+class RoundRobinAssigner extends TaskAssigner {
+  private var idx = 0
+
+  override def construct(workOffer: Seq[WorkerOffer]): Unit = {
+    offer = Random.shuffle(workOffer.map(o => new OfferState(o)))
+  }
+
+  override def init(): Unit = {
+    idx = 0
+  }
+
+  override def hasNext: Boolean = idx < offer.size
+
+  override def getNext(): OfferState = {
+    offer(idx)
+  }
+
+  override def offerAccepted(assigned: Boolean): Unit = {
+    idx += 1
+  }
+
+  override def reset(): Unit = {
+    super.reset
+    idx = 0
+  }
+}
+
+/**
+ * Assign the task to workers with the most available cores.
+ */
+class BalancedAssigner extends TaskAssigner {
+  private var maxHeap: PriorityQueue[OfferState] = _
+  private var currentOffer: OfferState = _
+
+  override def construct(workOffer: Seq[WorkerOffer]): Unit = {
+    offer = Random.shuffle(workOffer.map(o => new OfferState(o)))
+  }
+
+  implicit val ord: Ordering[OfferState] = new Ordering[OfferState] {
+    def compare(x: OfferState, y: OfferState): Int = {
+      return Ordering[Int].compare(x.coresAvailable, y.coresAvailable)
+    }
+  }
+
+  override def init(): Unit = {
+    maxHeap = new PriorityQueue[OfferState]()
+    offer.filter(_.coresAvailable >= CPUS_PER_TASK).foreach(maxHeap.enqueue(_))
+  }
+
+  override def hasNext: Boolean = maxHeap.nonEmpty
+
+  override def getNext(): OfferState = {
+    currentOffer = maxHeap.dequeue()
+    currentOffer
+  }
+
+  override def offerAccepted(assigned: Boolean): Unit = {
+    if (currentOffer.coresAvailable >= CPUS_PER_TASK && assigned) {
```

Comment:
If it is rejected, it is not valid for this round of assignment for this specific task set anymore. Because it means it is not valid for all tasks in the task set.